File: src/main/webapp/js/Integration/Observation.js
var Reference = require("./Reference");
var ValueQuantity = require("./ValueQuantity");
var Coding = require("./Coding");

function Observation(resourceType, id, code, valueQuantity,
		appliesDateTime, status, reliability, subject, encounter) {
	this.resourceType = resourceType;
	this.id = id;
	this.code = code;
	this.valueQuantity = valueQuantity;
	this.appliesDateTime = appliesDateTime;
	this.status = status;
	this.reliability = reliability;
	this.subject = subject;
	this.encounter = encounter;
};

Observation.prototype.prettyPrint = function() {
	return "Resource Type: " + this.resourceType
	+ " ID: " + this.id
	+ " Code: "+JSON.stringify(this.code);
	+ " Value Quantity: "+JSON.stringify(this.code);
	+ " Applies Date Time: "+this.appliesDateTime
	+ " Status: "+this.status
	+ " Reliability: "+this.reliability
	+ " Patient ID: "+this.subject
	+ " Encounter ID: "+this.encounter;
};

Observation.prototype.jsonPrint = function() {
	var slash = "/";
	return {	resourceType: this.resourceType,
		id: this.id,
		code: { coding: [this.code] },
		valueQuantity: this.valueQuantity,
		appliesDateTime: this.appliesDateTime,
		status: this.status,
		reliability: this.reliability,
		subject: new Reference("Patient" + slash + this.subject),
		encounter: new Reference(this.encounter)
	};
};

Observation.prototype.getObservationByText = function getObservationByText(text) {
	var observation = JSON.parse(text);
	var resourceType = observation.resourceType;
	var id = observation.id;
	var code = new Coding(observation.code.coding[0].system, observation.code.coding[0].code, observation.code.coding[0].display);
	var valueQuantity = new ValueQuantity(observation.valueQuantity.value, observation.valueQuantity.units, observation.valueQuantity.system, observation.valueQuantity.code);
	var appliesDateTime = observation.appliesDateTime;
	var status = observation.status;
	var reliability = observation.reliability;
	var subject = observation.subject.reference.replace("Patient/","");
	var encounter = observation.encounter.reference.replace("Encounter/","");
	return new Observation(resourceType, id, code, valueQuantity, appliesDateTime, status, reliability, subject, encounter);
};

module.exports = Observation;

File: src/main/java/Entities/Observation.java
package Entities;

import org.json.*;

public class Observation extends Entity {

	private static final String SLASH = "/";
	private static final String PATIENT = "Patient";
	private static final String ENCOUNTER = "encounter";
	private static final String SUBJECT = "subject";
	private static final String RELIABILITY = "reliability";
	private static final String STATUS = "status";
	private static final String APPLIES_DATE_TIME = "appliesDateTime";
	private static final String VALUE_QUANTITY = "valueQuantity";
	private static final String CODE = "code";
	private static final String CODING = "coding";

	public Observation(String resourceType, String id, Coding coding, ValueQuantity valueQuantity,
			String appliesDateTime, String status, String reliability, String subject, String encounterId) {
		super(resourceType, id, null);

		JSONObject codeJson = new JSONObject();
		JSONArray codeArray = new JSONArray();
		codeArray.put(coding.getJSONObject());
		codeJson.put(CODING, codeArray);
		entity.put(CODE, codeJson);
		
		entity.put(VALUE_QUANTITY, valueQuantity.getJSONObject());
		entity.put(APPLIES_DATE_TIME, appliesDateTime);
		entity.put(STATUS, status);
		entity.put(RELIABILITY, reliability);
		entity.put(SUBJECT, new Reference(PATIENT + SLASH + subject).getJSONObject());
		entity.put(ENCOUNTER, new Reference(encounterId).getJSONObject());
	}
	
	public Observation(String json) {
		super(null, null, json);
	}
}
